### PR TITLE
Parse and handle the stereoCare or STBOX flags in CTABs

### LINKED ISSUE #2916: Add _ctab, _mol2, _pdb to allow direct mol construction from strings
Analogous to what we have for `_smiles` and `_smarts`

### PR DESCRIPTION
**Pull request by greglandrum**
the `stereoCare` flags are now properly parsed on atoms and bonds and I've extended `AdjustQueryProperties()` to be able to interpret them.

Other things in here:
- Moves the function for populating `AdjustQueryParameters` from JSON into the public API (this used to be in the PostgreSQL cartridge code).
- Fixes #2916, which I should have done a long time ago
- A change to the way some of the catch2 tests are built to improve compilation speed

### CODE REVIEW

File: Code/GraphMol/FileParsers/testPropertyLists.cpp [hunk 1]
```diff
@@ -7,8 +7,6 @@
 //  of the RDKit source tree.
 //
 
-#define CATCH_CONFIG_MAIN  // This tells Catch to provide a main() - only do
```
Comment (greglandrum) [inline Code/GraphMol/FileParsers/testPropertyLists.cpp]: Unfortunately this file ended up with a bunch of formatting changes.
The only substantive change here is the removal of this `#define`

File: Code/GraphMol/catch_graphmol.cpp [hunk 2]
```diff
@@ -1108,3 +1117,248 @@ TEST_CASE("github #2895: acepentalene aromaticity perception ",
...
+M  END
+)CTAB"_ctab;
+    REQUIRE(mol);
+    REQUIRE(mol->getBondBetweenAtoms(0, 1));
+    CHECK(mol->getBondBetweenAtoms(0, 1)->getStereo() ==
+          Bond::BondStereo::STEREOE);
+    MolOps::AdjustQueryParameters ps;
+    ps.useStereoCareForBonds = true;
+    MolOps::adjustQueryProperties(*mol, &ps);
+    CHECK(mol->getBondBetweenAtoms(0, 1)->getStereo() ==
```
Comment (bp-kelley) [inline Code/GraphMol/catch_graphmol.cpp]: Could you also have a test for useStereoCareForBonds = false?  It's hard to tell in this test if anything actually happened.

Comment (bp-kelley) [reply Code/GraphMol/catch_graphmol.cpp]: Comment so I can start a  review

File: Code/GraphMol/catch_graphmol.cpp [hunk 3]
```diff
@@ -1108,3 +1117,248 @@ TEST_CASE("github #2895: acepentalene aromaticity perception ",
...
```
Comment (bp-kelley) [inline Code/GraphMol/catch_graphmol.cpp]: Here we go :)  This does the test I want :)

File: Code/PgSQL/rdkit/expected/rdkit-91.out [hunk 4]
```diff
@@ -922,7 +922,7 @@ select 'C1C(C)C1CCCC'::mol @> mol_adjust_query_properties('C1CC1CC'::mol,'{"adju
 (1 row)
 
 select 'C1C(C)C1CCCC'::mol @> mol_adjust_query_properties('C1CC1CC'::mol,'{"adjustDegreeFlags":"bogus"}');
-ERROR:  bad which string provided 'BOGUS'
+ERROR:  Unknown flag value: 'BOGUS'. Flags ignored.
```
Comment (bp-kelley) [inline Code/PgSQL/rdkit/expected/rdkit-91.out]: This implies that all flags are ignored which I think is not true.  Only *this* flag is ignored.

File: Code/GraphMol/catch_graphmol.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/catch_graphmol.cpp]: `useStereoCareForBonds=false` is the default, so basically all of the other tests use that.

File: Code/PgSQL/rdkit/expected/rdkit-91.out (hunk 4 above)
Comment (greglandrum) [reply Code/PgSQL/rdkit/expected/rdkit-91.out]: If any of the `adjustDegreeFlags` are bad, then they are all ignored.

Comment (bp-kelley) [reply Code/PgSQL/rdkit/expected/rdkit-91.out]: All right, I rescind my comment.

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp [hunk 5]
```diff
@@ -869,4 +865,175 @@ M  END)CTAB";
...
+    4     2     6  1
+    5     2     1  1
+)DATA"_mol2;
+    REQUIRE(!mol);
+  }
+}
+
+TEST_CASE("handling STBOX properties from v3k ctabs", "[feature,v3k]") {
+  SECTION("atoms and bonds") {
+    auto mol = R"CTAB(basic test
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/file_parsers_catch.cpp]: handy.

File: Code/GraphMol/catch_graphmol.cpp [hunk 6]
```diff
@@ -1108,3 +1117,248 @@ TEST_CASE("github #2895: acepentalene aromaticity perception ",
...
+          Bond::BondStereo::STEREONONE);
+  }
```
Comment (d-b-w) [inline Code/GraphMol/catch_graphmol.cpp]: Would it be useful to add a test here demonstrating what would happen if you fed in a mol that came from SMILES or something?

    auto mol = "C/C=C/F"_smiles;
    ...
    ps.useStereoCareForBonds = true;
    MolOps::adjustQueryProperties(*mol, &ps);
    CHECK(b->getStereo() == Bond::BondStereo::STEREONONE);

Comment (greglandrum) [reply Code/GraphMol/catch_graphmol.cpp]: added

File: Code/GraphMol/FileParsers/file_parsers_catch.cpp [hunk 7]
```diff
@@ -869,4 +865,175 @@ M  END)CTAB";
...
+M  V30 END ATOM
+M  V30 BEGIN BOND
+M  V30 1 1 2 3
+M  V30 2 1 1 4
+M  V30 3 2 1 2 STBOX=1
+M  V30 END BOND
+M  V30 END CTAB
+M  END
+)CTAB"_ctab;
+    REQUIRE(mol);
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/file_parsers_catch.cpp]: To start by answering the last question: this PR doesn't change the behavior of substructure matching at all. If you do a substructure match that uses chirality info, and search with a bond that has specified stereochemistry, that stereochemistry info will be used when doing the query.

The changes here are being made in `MolOps::adjustQueryProperties()`, which is something that would normally be called by a system to prepare the query molecule before actually executing a query. At that point in the process the developer generally knows where the query came from (SMILES, SMARTS, MOL, etc) and can make an informed decision about which properties need to be adjusted to get the query they want. You generally would not use the `useStereoCareForBonds` option when dealing with things from SMILES/SMARTS.

File: Code/GraphMol/MolOps.h [hunk 8]
```diff
@@ -310,40 +310,44 @@ typedef enum {
...
+
+  bool adjustHeavyDegree = false; /**< adjust the heavy-atom degree instead of
+                               overall degree */
+  std::uint32_t adjustHeavyDegreeFlags =
+      ADJUST_IGNOREDUMMIES | ADJUST_IGNORECHAINS;
+
+  bool adjustRingChain = false; /**< add ring-chain queries */
+  std::uint32_t adjustRingChainFlags = ADJUST_IGNORENONE;
+
+  bool useStereoCareForBonds =
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolOps.h]: I like the idea of modifying the parser to move the property to the bond when both atoms have it set. That simplifies the logic in this code and more accurately reflects the intent.

[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/MolOps.h]: > I still have some concern here that this puts the requirement to know about the source of the data at the point that the SMARTS query is done.

This PR doesn't change anything that happens at query time. It's all about the preprocessing that you choose to do to the query molecule before running the query.

File: Code/GraphMol/AdjustQuery.cpp [hunk 9]
```diff
@@ -285,16 +285,9 @@ void adjustQueryProperties(RWMol &mol, const AdjustQueryParameters *inParams) {
           bool preserve = false;
           int val = 0;
           // is stereoCare set on the bond or both atoms?
-          if (bnd->getPropIfPresent(common_properties::molStereoCare, val) ||
+          if (bnd->getPropIfPresent(common_properties::molStereoCare, val) &&
```
Comment (bp-kelley) [inline Code/GraphMol/AdjustQuery.cpp]: Interestingly, these are functionally equivalent :)

**Comment by bp-kelley** [comment]
One thing I may add is a stereo care namespace, I see a lot of code like:

```
setProp(common_properties::molStereoCare, 1)
```

Where it might be nicer to do:

```
setProp(common_properties::molStereoCare, MolStereoCare.STEREO_MUST_MATCH);
```

Other than that, I'm happy to merge.

File: Code/GraphMol/FileParsers/MolFileParser.cpp [hunk 10]
```diff
@@ -2319,6 +2331,20 @@ void ParseV3000BondBlock(std::istream *inStream, unsigned int &line,
...
+    // set the stereoCare property on the bond if it's not set already and both
+    // the beginning and end atoms have it set:
+    int care1 = 0;
+    int care2 = 0;
+    if (!bond->hasProp(common_properties::molStereoCare) &&
+        mol->getAtomWithIdx(bond->getBeginAtomIdx())
+            ->getPropIfPresent(common_properties::molStereoCare, care1) &&
+        mol->getAtomWithIdx(bond->getEndAtomIdx())
+            ->getPropIfPresent(common_properties::molStereoCare, care2)) {
+      if (care1 == care2) {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/FileParsers/MolFileParser.cpp]: This is true when the property comes from V3000 - where the file has to specifically include `STBOX=0` - but not for V2000 (where "0" or missing is just the default and we don't set it).
I figured it makes sense to go ahead and set the property here and then to check for the value when processing it in `adjustQueryProperties()` (this is what should currently be happening), but I'm not wedded to that. We could just as easily not set it here if the value is `0`

**Comment by bp-kelley** [comment]
@greglandrum @d-b-w I'm happy to merge.  Any objection?  Then we can poke #2829

**Comment by greglandrum** [comment]
Yep, LGTM

**Pull request merged**